javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +95,1381 @@ class Formatter(
-        startIndex += 1
-      return padding(result, startIndex)
-    }
+      // Insert grouping separators, if required
+      val rest =
+        if (flags.useGroupingSeps) insertGroupingCommas(localeInfo, rest0)
+        else rest0
 
-    private def wrapParentheses(result: StringBuilder): StringBuilder = {
-      // delete the '-'
-      result.deleteCharAt(0)
-      result.insert(0, '(')
-      if (formatToken.isFlagSet(FormatToken.FLAG_ZERO)) {
-        formatToken.setWidth(formatToken.getWidth() - 1)
-        padding(result, 1)
-        result.append(')')
-      } else {
-        result.append(')')
-        padding(result, 0)
-      }
-      result
+      // Apply uppercase, localization, pad and send
+      padAndSendToDest(
+        localeInfo,
+        flags,
+        width,
+        prefix,
+        localeInfo.localizeNumber(applyNumberUpperCase(flags, rest)))
     }
+  }
 
-    private def transformFromSpecialNumber(): String = {
-      var source: String = null
-
-      if (!(arg.isInstanceOf[Number]) || arg.isInstanceOf[BigDecimal])
-        return null
-
-      val number = arg.asInstanceOf[Number]
-      val d      = number.doubleValue()
-      if (java.lang.Double.isNaN(d))
-        source = "NaN"
-      else if (java.lang.Double.isInfinite(d)) {
-        if (d >= 0) {
-          if (formatToken.isFlagSet(FormatToken.FLAG_ADD))
-            source = "+Infinity"
-          else if (formatToken.isFlagSet(FormatToken.FLAG_SPACE))
-            source = " Infinity"
-          else
-            source = "Infinity"
-        } else {
-          if (formatToken.isFlagSet(FormatToken.FLAG_PARENTHESIS))
-            source = "(Infinity)"
-          else
-            source = "-Infinity"
-        }
-      }
-
-      if (null != source) {
-        formatToken.setPrecision(FormatToken.UNSET)
-        formatToken.setFlags(formatToken.getFlags() & (~FormatToken.FLAG_ZERO))
-        source = padding(new StringBuilder(source), 0)
+  /** Inserts grouping commas at the right positions for the locale.
+   *
+   *  We already insert the ',' character, regardless of the locale. That is
+   *  fixed later by `localeInfo.localizeNumber`. The only locale-sensitive
+   *  behavior in this method is the grouping size.
+   *
+   *  The reason is that we do not want to insert a character that would
+   *  collide with another meaning (such as '.') at this point.
+   */
+  private def insertGroupingCommas(localeInfo: LocaleInfo,
+                                   s: String): String = {
+    val groupingSize = localeInfo.groupingSize
+
+    val len   = s.length
+    var index = 0
+    while (index != len && { val c = s.charAt(index); c >= '0' && c <= '9' }) {
+      index += 1
+    }
+
+    index -= groupingSize
+
+    if (index <= 0) {
+      s
+    } else {
+      var result = s.substring(index)
+      while (index > groupingSize) {
+        val next = index - groupingSize
+        result = s.substring(next, index) + groupingSeparator + result
```

Comment:
Please use a hard-coded `","` here, since the documentation of the function specifically says that it always inserts commas.